### PR TITLE
cl-context: build code independently with extension function

### PR DESCRIPTION
**Pull request by liuyinhangx**

### CODE REVIEW

File: modules/ocl/cl_context.cpp [hunk 1]
```diff
@@ -423,10 +434,15 @@ CLContext::import_dma_buffer (const cl_import_buffer_info_intel &import_info)
...
     cl_mem mem_id = NULL;
     cl_int errcode = CL_SUCCESS;
-    if (!is_valid())
+    cl_platform_id platform_id = _device->get_platform_id ();
+    if (!is_valid() || !platform_id)
         return NULL;
 
-    mem_id = clCreateBufferFromFdINTEL (_context_id, &import_info, &errcode);
+    OCLCREATEBUFFERFROMFDINTEL *oclCreateBufferFromFdINTEL =
+        (OCLCREATEBUFFERFROMFDINTEL *) clGetExtensionFunctionAddressForPlatform (platform_id, "clCreateBufferFromFdINTEL");
```
Comment (windyuan) [inline modules/ocl/cl_context.cpp]: move clGetExtensionFunctionAddressForPlatform  to cl_device.cpp.
add interface void *CLDevice::get_extension_function(const char* func_name);

File: modules/ocl/cl_device.h [hunk 2]
```diff
@@ -58,6 +58,9 @@ class CLDevice {
     const CLDevieInfo &get_device_info () {
         return _device_info;
     }
+    cl_platform_id get_platform_id () {
+        return _platform_id;
+    }
```
Comment (windyuan) [inline modules/ocl/cl_device.h]: no need to export platform_id

File: modules/ocl/cl_context.cpp [hunk 3]
```diff
@@ -441,10 +457,15 @@ CLContext::create_va_image (const cl_libva_image &image_info)
...
     cl_mem mem_id = NULL;
     cl_int errcode = CL_SUCCESS;
-    if (!is_valid())
+    cl_platform_id platform_id = _device->get_platform_id ();
+    if (!is_valid() || !platform_id)
         return NULL;
 
-    mem_id = clCreateImageFromLibvaIntel (_context_id, &image_info, &errcode);
+    OCLCREATEIMAGEFROMLIBVAINTEL *oclCreateImageFromLibvaIntel =
+        (OCLCREATEIMAGEFROMLIBVAINTEL *) clGetExtensionFunctionAddressForPlatform (platform_id, "clCreateImageFromLibvaIntel");
```
Comment (windyuan) [inline modules/ocl/cl_context.cpp]: OCLCREATEIMAGEFROMLIBVAINTEL defined for function pointer. no need to define another pointer.
just OCLCREATEIMAGEFROMLIBVAINTEL oclCreateImageFromLibvaIntel

File: modules/ocl/cl_context.cpp [hunk 4]
```diff
@@ -458,10 +479,15 @@ CLContext::import_dma_image (const cl_import_image_info_intel &import_info)
...
     cl_int errcode = CL_SUCCESS;
-    if (!is_valid())
+    cl_platform_id platform_id = _device->get_platform_id ();
+    if (!is_valid() || !platform_id)
         return NULL;
 
-    mem_id = clCreateImageFromFdINTEL (_context_id, &import_info, &errcode);
+    OCLCREATEIMAGEFROMFDINTEL *oclCreateImageFromFdINTEL =
+        (OCLCREATEIMAGEFROMFDINTEL *) clGetExtensionFunctionAddressForPlatform (platform_id, "clCreateImageFromFdINTEL");
+    XCAM_ASSERT (oclCreateImageFromFdINTEL);
```
Comment (windyuan) [inline modules/ocl/cl_context.cpp]: All extension function name should be defined with Macros. e.g
        #define OCL_EXT_NAME_CREATE_IMAGE_FROM_FD_INTEL "clCreateImageFromFdINTEL"

**Comment by liuyinhangx** [comment]
Update patch, pls review.

**Comment by windyuan** [comment]
merged. thanks.
commit 0305ba5f938d84cfb8debde8edef56cbebbfd9e2
some changes.

      diff --git a/modules/ocl/cl_context.cpp b/modules/ocl/cl_context.cpp
      index ee97a8f..0b5273b 100644
      --- a/modules/ocl/cl_context.cpp
      +++ b/modules/ocl/cl_context.cpp
      @@ -417,7 +417,7 @@ CLContext::create_va_buffer (uint32_t bo_name)
      
           clCreateBufferFromLibvaIntel_fn oclCreateBufferFromLibvaIntel =
               (clCreateBufferFromLibvaIntel_fn) _device->get_extension_function (OCL_EXT_NAME_CREATE_BUFFER_FROM_LIBVA_INTEL);
      -    XCAM_ASSERT (oclCreateBufferFromLibvaIntel);
      +    XCAM_FAIL_RETURN(ERROR, oclCreateBufferFromLibvaIntel, NULL, "create buffer failed since extension was not found");
      
           mem_id = oclCreateBufferFromLibvaIntel (_context_id, bo_name, &errcode);
           XCAM_FAIL_RETURN(
      
      .............
      
      diff --git a/modules/ocl/cl_device.cpp b/modules/ocl/cl_device.cpp
      index 694d6e4..58b07d8 100644
      --- a/modules/ocl/cl_device.cpp
      +++ b/modules/ocl/cl_device.cpp
      @@ -72,13 +72,13 @@ CLDevice::get_extension_function (const char *func_name)
           XCAM_ASSERT (func_name);
           void *ext_func = NULL;
      
      -#if CL_VERSION_1_2 == 1
      +#if defined (CL_VERSION_1_2) && (CL_VERSION_1_2 == 1)
           ext_func = (void *) clGetExtensionFunctionAddressForPlatform (_platform_id, func_name);
       #else
           ext_func = (void *) clGetExtensionFunctionAddress (func_name);
       #endif

**Pull request closed without merging**